cmd/docker-proxy/main.go

```diff
@@ -31,41 +47,114 @@ func main() {
+		}
+		hostAddr := &sctp.SCTPAddr{IPAddrs: []net.IPAddr{{IP: config.HostIP}}, Port: config.HostPort}
+		container := &sctp.SCTPAddr{IPAddrs: []net.IPAddr{{IP: config.ContainerIP}}, Port: config.ContainerPort}
+		listener, err = sctp.ListenSCTP("sctp"+string(ipv), hostAddr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to listen on %s: %w", hostAddr, err)
+		}
+		p, err = NewSCTPProxy(listener, container)
+	default:
+		return nil, fmt.Errorf("unsupported protocol %s", config.Proto)
+	}
+
+	return p, err
+}
+
+type ProxyConfig struct {
+	Proto                   string
+	HostIP, ContainerIP     net.IP
+	HostPort, ContainerPort int
+	ListenSock              *os.File
+}
+
 // parseFlags parses the flags passed on reexec to create the TCP/UDP/SCTP
 // net.Addrs to map the host and container ports.
-func parseFlags() (host net.Addr, container net.Addr) {
+func parseFlags() ProxyConfig {
 	var (
-		proto         = flag.String("proto", "tcp", "proxy protocol")
-		hostIP        = flag.String("host-ip", "", "host ip")
-		hostPort      = flag.Int("host-port", -1, "host port")
-		containerIP   = flag.String("container-ip", "", "container ip")
-		containerPort = flag.Int("container-port", -1, "container port")
-		printVer      = flag.Bool("v", false, "print version information and quit")
-		printVersion  = flag.Bool("version", false, "print version information and quit")
+		config      ProxyConfig
+		useListenFd bool
+		printVer    bool
 	)
-
+	flag.StringVar(&config.Proto, "proto", "tcp", "proxy protocol")
+	flag.TextVar(&config.HostIP, "host-ip", net.IPv4zero, "host ip")
+	flag.IntVar(&config.HostPort, "host-port", -1, "host port")
+	flag.TextVar(&config.ContainerIP, "container-ip", net.IPv4zero, "container ip")
+	flag.IntVar(&config.ContainerPort, "container-port", -1, "container port")
+	flag.BoolVar(&useListenFd, "use-listen-fd", false, "use a supplied listen fd")
+	flag.BoolVar(&printVer, "v", false, "print version information and quit")
+	flag.BoolVar(&printVer, "version", false, "print version information and quit")
 	flag.Parse()
 
-	if *printVer || *printVersion {
+	if printVer {
 		fmt.Printf("docker-proxy (commit %s) version %s\n", dockerversion.GitCommit, dockerversion.Version)
 		os.Exit(0)
 	}
 
-	switch *proto {
-	case "tcp":
-		host = &net.TCPAddr{IP: net.ParseIP(*hostIP), Port: *hostPort}
-		container = &net.TCPAddr{IP: net.ParseIP(*containerIP), Port: *containerPort}
-	case "udp":
-		host = &net.UDPAddr{IP: net.ParseIP(*hostIP), Port: *hostPort}
-		container = &net.UDPAddr{IP: net.ParseIP(*containerIP), Port: *containerPort}
-	case "sctp":
-		host = &sctp.SCTPAddr{IPAddrs: []net.IPAddr{{IP: net.ParseIP(*hostIP)}}, Port: *hostPort}
-		container = &sctp.SCTPAddr{IPAddrs: []net.IPAddr{{IP: net.ParseIP(*containerIP)}}, Port: *containerPort}
-	default:
-		log.Fatalf("unsupported protocol %s", *proto)
+	if useListenFd {
+		config.ListenSock = os.NewFile(listenSockFd, "listen-sock")
```

Comment:
Done. I guess it'll be more significant if we make `docker-proxy` handle more than one connection.

Reply:
`docker-proxy` would not be fit for purpose if only one client could connect to the container port at a time. It definitely is able to concurrently proxy multiple connections. https://github.com/moby/moby/blob/27b322e56bc04da8e431fe792c742f7b52df9c46/cmd/docker-proxy/tcp_proxy.go#L77-L84

Reply:
Oh right, of course ... I was thinking about the one listener vs proxying multiple ports.

Reply:
The point being, the per-connection / traffic carrying sockets returned by Accept were already nonblocking and there's only one listener.

Reply:
Right, right; only one thread would be blocking on the `accept` syscall per listener. Fair point. Still, saving a thread per non-idle proxy process at the cost of one line of code with no ongoing maintenance or cognitive-complexity costs is well worth doing. And arguably it's avoiding a resource-usage regression rather than adding a optimization.

Reply:
Yes, it was worth sorting out - thank you for spotting it.